<a href="https://colab.research.google.com/github/askot19/Portfolio/blob/main/Churn%20Prediction/1%20-%20Explora%C3%A7%C3%A3o%20e%20Limpeza%20de%20Dados/01%20-%20Limpeza%20dos%20Dados.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<p align = 'left'>
<img src = 'https://i.imgur.com/8LTNXxF.jpg'>
</p>

---
## **Data Scientist & Analyst**

*by [Paulo Henrique de Melo Ferreira](https://www.linkedin.com/in/paulo-henrique-ferreira-20151374/)*

---

## Predição de Churn

Neste projeto, eu construirei um modelo de machine learning para predizer o Churn de uma empresa de telefonia. Ele envolve aprendizado supervisionado (utilizando um dataset) para classificação, onde a evasão positiva é determinada como 1, e a não evasão é 0.

Utilizarei o pipeline baseado na metodologia CRISP-DM:

1. Definir o problema de negócio.
2. Coleta de dados e entendimento de seus significados.
3. Limpeza de dados.
4. Exploração dos dados.
5. Divisão dos dados para treinamento e predição.
6. Treinamento de modelo, comparação, seleção de características e ajustes.
7. Produção final do modelo e avaliação.
8. Conclusões e interpretação dos resultados obtidos do modelo.

Neste notebook, farei a Limpeza de Dados, cobrindo os passos de 1 a 3 do pipeline acima. O principal objetivo aqui é tratar os dados que nos darão informação relevante sobre os padrões de Churn dentro das características disponibilizadas. Dessa maneira, mesmo antes de se construir um modelo, será possível identificar os perfis de Churn sofridos pela empresa. Ademais, irei detalhar esses passos abaixo, explicando o porquê de cada tomada de decisão.

1. Problema de Negócio

Um gerente de uma empresa de serviços telefônicos está preocupado com o crescente número de clientes evadidos. Seria interessante realizar uma predição para eles do quão suscetível à evasão um cliente possa ser, para que eles possam agir proativamente e fornecer melhores serviços, ações de marketing e alterar a decisão desses clientes.

1.1 Qual o contexto?

Em tempos atuais empresas de telefonia oferecem não apenas serviços voltados para comunicação, para manterem-se ativos no mercado são fornecidos também atividades voltadas para lazer, banda larga, backups e segurança online.
Com o auxílio de análises estatísticas, identificaremos quais os Indicadores Chaves de Performance (KPIs) do Churn de clientes desta empresa, podendo traçar estratégias e entender o que motiva ou não a evasão.

- Quais são os objetivos do projeto?

1. Identificar os fatores associados com o Churn de clientes;
2. Construir um modelo capaz de prever de maneira acurada a probabilidade de um cliente desistir dos serviços;
3. Oferecer planos de ação para reduzir a evasão de clientes.

- Quais são os benefícios?

1. Redução de custos;
2. Melhorar a retenção de clientes;
3. Melhorar a experiência dos clientes;
4. Marketing direcionado;
5. Proteção da receita.

E como resultado, o problema de negócio será resolvido.

## Importando as bibliotecas

In [1]:
# Carregando pacostes para análise de dados
import pandas as pd
import numpy as np

In [2]:
# Aqui carregamos os dados por meio de um arquivo no formato JSON, disponibilizado pelo github
dados = pd.read_json('https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json')
dados.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [3]:
# Como o carregamento para o Data Frame não ocorreu como o esperado, vamos explorar o que cada campo armazena em sua primeira linha.
dados.customer[0]
# Aqui observamos que são dados gerais referentes ao cliente

{'gender': 'Female',
 'SeniorCitizen': 0,
 'Partner': 'Yes',
 'Dependents': 'Yes',
 'tenure': 9}

In [4]:
# Dados acerca da linha do cliente
dados.phone[0]

{'PhoneService': 'Yes', 'MultipleLines': 'No'}

In [5]:
# Dados a respeito de serviços de internet
dados.internet[0]

{'InternetService': 'DSL',
 'OnlineSecurity': 'No',
 'OnlineBackup': 'Yes',
 'DeviceProtection': 'No',
 'TechSupport': 'Yes',
 'StreamingTV': 'Yes',
 'StreamingMovies': 'No'}

In [6]:
# Dados acerca do contrato e formas de pagamento
dados.account[0]

{'Contract': 'One year',
 'PaperlessBilling': 'Yes',
 'PaymentMethod': 'Mailed check',
 'Charges': {'Monthly': 65.6, 'Total': '593.3'}}

In [7]:
# A partir daqui vamos normalizar os dados, a fim de poder trabalhar melhor com o Data Frame
customer_dados = pd.json_normalize(dados.customer)

In [8]:
phone_dados = pd.json_normalize(dados.phone)
phone_dados.head()

PhoneService MultipleLines
0          Yes            No
1          Yes           Yes
2          Yes            No
3          Yes            No
4          Yes            No

In [9]:
internet_dados = pd.json_normalize(dados.internet)
internet_dados.head()

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No         Yes   
1             DSL             No           No               No          No   
2     Fiber optic             No           No              Yes          No   
3     Fiber optic             No          Yes              Yes          No   
4     Fiber optic             No           No               No         Yes   

  StreamingTV StreamingMovies  
0         Yes              No  
1          No             Yes  
2          No              No  
3         Yes             Yes  
4         Yes              No

In [10]:
account_dados = pd.json_normalize(dados.account)
account_dados.head()

Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0        One year              Yes      Mailed check             65.6   
1  Month-to-month               No      Mailed check             59.9   
2  Month-to-month              Yes  Electronic check             73.9   
3  Month-to-month              Yes  Electronic check             98.0   
4  Month-to-month              Yes      Mailed check             83.9   

  Charges.Total  
0         593.3  
1         542.4  
2        280.85  
3       1237.85  
4         267.4

In [11]:
# Definindo uma função para concatenar todas as tabelas ao longo das colunas
def normalize_json(dataframe):
    return_dataframe = pd.DataFrame()
    for column in list(dados.columns[2:]):
        dataframe_column = pd.json_normalize(dataframe[column])
        return_dataframe = pd.concat([return_dataframe, dataframe_column], axis=1)

    return pd.concat([dataframe[list(dados.columns[:2])], return_dataframe], axis=1)

In [12]:
# Aplicando a função e rearmazenando os dados na respectiva variável, vamos também visualizar o resultado
dados = normalize_json(dados)
dados

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7267 rows x 21 columns]

Agora com os dados devidamente armazenados em um Data Frame, podemos iniciar a limpeza de dados deste. Mas antes, vamos entender do que se tratam cada uma das variáveis de análise.

In [13]:
# Consultando os nomes dos das colunas no Data Frame
dados.columns.values

array(['customerID', 'Churn', 'gender', 'SeniorCitizen', 'Partner',
       'Dependents', 'tenure', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'Charges.Monthly', 'Charges.Total'], dtype=object)

Para facilitar o entendimento das colunas, vamos traduzi-las para conferir maior acessibilidade à equipe de trabalho.

In [14]:
nomes = {'customerID': 'ID', 'Churn': 'Evasao', 'gender': 'Genero', 'SeniorCitizen': 'Senior', 'Partner': 'Casado(a)',
       'Dependents': 'Dependentes', 'tenure': 'Posse', 'PhoneService': 'ServicoTelefonico', 'MultipleLines': 'MultiplasLinhas',
       'InternetService': 'ServicodeInternet', 'OnlineSecurity': 'SegurancaOnline', 'OnlineBackup': 'BackupOnline',
       'DeviceProtection': 'ProtecaodeDispositivo', 'TechSupport': 'SuporteTecnico', 'StreamingTV': 'StreamingTV',
       'StreamingMovies': 'StreamingFilmes', 'Contract': 'Contrato', 'PaperlessBilling': 'CobrancasemPapel', 'PaymentMethod': 'MetododePagamento',
       'Charges.Monthly': 'CobrancasMensais', 'Charges.Total': 'CobrancasTotais'}

In [15]:
dados.rename(columns = nomes, inplace = True)
dados

ID Evasao  Genero  Senior Casado(a) Dependentes  Posse  \
0     0002-ORFBO     No  Female       0       Yes         Yes      9   
1     0003-MKNFE     No    Male       0        No          No      9   
2     0004-TLHLJ    Yes    Male       0        No          No      4   
3     0011-IGKFF    Yes    Male       1       Yes          No     13   
4     0013-EXCHZ    Yes  Female       1       Yes          No      3   
...          ...    ...     ...     ...       ...         ...    ...   
7262  9987-LUTYD     No  Female       0        No          No     13   
7263  9992-RRAMN    Yes    Male       0       Yes          No     22   
7264  9992-UJOEL     No    Male       0        No          No      2   
7265  9993-LHIEB     No    Male       0       Yes         Yes     67   
7266  9995-HOTOH     No    Male       0       Yes         Yes     63   

     ServicoTelefonico   MultiplasLinhas ServicodeInternet  ... BackupOnline  \
0                  Yes                No               DSL  ...          Yes   
1                  Yes               Yes               DSL  ...           No   
2                  Yes                No       Fiber optic  ...           No   
3                  Yes                No       Fiber optic  ...          Yes   
4                  Yes                No       Fiber optic  ...           No   
...                ...               ...               ...  ...          ...   
7262               Yes                No               DSL  ...           No   
7263               Yes               Yes       Fiber optic  ...           No   
7264               Yes                No               DSL  ...          Yes   
7265               Yes                No               DSL  ...           No   
7266                No  No phone service               DSL  ...          Yes   

     ProtecaodeDispositivo SuporteTecnico StreamingTV StreamingFilmes  \
0                       No            Yes         Yes              No   
1                       No             No          No             Yes   
2                      Yes             No          No              No   
3                      Yes             No         Yes             Yes   
4                       No            Yes         Yes              No   
...                    ...            ...         ...             ...   
7262                    No            Yes          No              No   
7263                    No             No          No             Yes   
7264                    No             No          No              No   
7265                   Yes            Yes          No             Yes   
7266                   Yes             No         Yes             Yes   

            Contrato CobrancasemPapel MetododePagamento CobrancasMensais  \
0           One year              Yes      Mailed check            65.60   
1     Month-to-month               No      Mailed check            59.90   
2     Month-to-month              Yes  Electronic check            73.90   
3     Month-to-month              Yes  Electronic check            98.00   
4     Month-to-month              Yes      Mailed check            83.90   
...              ...              ...               ...              ...   
7262        One year               No      Mailed check            55.15   
7263  Month-to-month              Yes  Electronic check            85.10   
7264  Month-to-month              Yes      Mailed check            50.30   
7265        Two year               No      Mailed check            67.85   
7266        Two year               No  Electronic check            59.00   

      CobrancasTotais  
0               593.3  
1               542.4  
2              280.85  
3             1237.85  
4               267.4  
...               ...  
7262            742.9  
7263           1873.7  
7264            92.75  
7265          4627.65  
7266           3707.6  

[7267 rows x 21 columns]

In [16]:
# Verificando valores únicos das colunas
dados['MultiplasLinhas'].unique()

array(['No', 'Yes', 'No phone service'], dtype=object)

In [17]:
dados['ServicodeInternet'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [18]:
dados['Contrato'].unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [19]:
dados['SegurancaOnline'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [20]:
dados['MetododePagamento'].unique()

array(['Mailed check', 'Electronic check', 'Credit card (automatic)',
       'Bank transfer (automatic)'], dtype=object)

In [21]:
dados['Senior'].unique()

array([0, 1])

In [22]:
# Traduzindo valores contidos nas colunas
dados.replace({'Yes': 'Sim', 'No': 'Nao', 'Female': 'Mulher', 'Male': 'Homem', 'No phone service': 'Sem servico telefonico', 'Fiber optic': 'Fibra optica', 'One year': 'Anual', 'Month-to-month': 'Mensal', 'Two year': 'Dois anos', 'No internet service': 'Sem servico de Internet', 'Mailed check': 'Correspondencia', 'Electronic check': 'Eletronico', 'Credit card (automatic)': 'Cartao de credito', 'Bank transfer (automatic)': 'Debito automatico'}, inplace = True)
dados

ID Evasao  Genero  Senior Casado(a) Dependentes  Posse  \
0     0002-ORFBO    Nao  Mulher       0       Sim         Sim      9   
1     0003-MKNFE    Nao   Homem       0       Nao         Nao      9   
2     0004-TLHLJ    Sim   Homem       0       Nao         Nao      4   
3     0011-IGKFF    Sim   Homem       1       Sim         Nao     13   
4     0013-EXCHZ    Sim  Mulher       1       Sim         Nao      3   
...          ...    ...     ...     ...       ...         ...    ...   
7262  9987-LUTYD    Nao  Mulher       0       Nao         Nao     13   
7263  9992-RRAMN    Sim   Homem       0       Sim         Nao     22   
7264  9992-UJOEL    Nao   Homem       0       Nao         Nao      2   
7265  9993-LHIEB    Nao   Homem       0       Sim         Sim     67   
7266  9995-HOTOH    Nao   Homem       0       Sim         Sim     63   

     ServicoTelefonico         MultiplasLinhas ServicodeInternet  ...  \
0                  Sim                     Nao               DSL  ...   
1                  Sim                     Sim               DSL  ...   
2                  Sim                     Nao      Fibra optica  ...   
3                  Sim                     Nao      Fibra optica  ...   
4                  Sim                     Nao      Fibra optica  ...   
...                ...                     ...               ...  ...   
7262               Sim                     Nao               DSL  ...   
7263               Sim                     Sim      Fibra optica  ...   
7264               Sim                     Nao               DSL  ...   
7265               Sim                     Nao               DSL  ...   
7266               Nao  Sem servico telefonico               DSL  ...   

     BackupOnline ProtecaodeDispositivo SuporteTecnico StreamingTV  \
0             Sim                   Nao            Sim         Sim   
1             Nao                   Nao            Nao         Nao   
2             Nao                   Sim            Nao         Nao   
3             Sim                   Sim            Nao         Sim   
4             Nao                   Nao            Sim         Sim   
...           ...                   ...            ...         ...   
7262          Nao                   Nao            Sim         Nao   
7263          Nao                   Nao            Nao         Nao   
7264          Sim                   Nao            Nao         Nao   
7265          Nao                   Sim            Sim         Nao   
7266          Sim                   Sim            Nao         Sim   

     StreamingFilmes   Contrato CobrancasemPapel MetododePagamento  \
0                Nao      Anual              Sim   Correspondencia   
1                Sim     Mensal              Nao   Correspondencia   
2                Nao     Mensal              Sim        Eletronico   
3                Sim     Mensal              Sim        Eletronico   
4                Nao     Mensal              Sim   Correspondencia   
...              ...        ...              ...               ...   
7262             Nao      Anual              Nao   Correspondencia   
7263             Sim     Mensal              Sim        Eletronico   
7264             Nao     Mensal              Sim   Correspondencia   
7265             Sim  Dois anos              Nao   Correspondencia   
7266             Sim  Dois anos              Nao        Eletronico   

     CobrancasMensais  CobrancasTotais  
0               65.60            593.3  
1               59.90            542.4  
2               73.90           280.85  
3               98.00          1237.85  
4               83.90            267.4  
...               ...              ...  
7262            55.15            742.9  
7263            85.10           1873.7  
7264            50.30            92.75  
7265            67.85          4627.65  
7266            59.00           3707.6  

[7267 rows x 21 columns]

# Dicionário de Dados

Com a tradução destes dados podemos agora fazer um entendimento de cada característica disponibilizada pela empresa:

* **``ID``**: Código de identificação do cliente;
* **``Evasao``**: Identificação se o cliente foi evadido (1) ou não (0);
* **``Genero``**: Identificação do sexo do cliente, masculino ou feminino;
* **``Senior``**: Classifica se o cliente está há mais de dois anos recebendo serviços;
* **``Casado(a)``**: Categoria em que se possui ou não cônjuge ;
* **``Dependentes``**: Se possui filhos;
* **``Posse``**: Há quanto tempo possui os serviços;
* **``ServicoTelefonico``**: Se possui serviço telefônico;
* **``MultiplasLinhas``**: Se possui mais de uma linha telefônica;
* **``ServicodeInternet``**: Quais os serviços de internet possui, podendo ser classificado em Fibra Óptica, Banda Larga ou sem serviço;
* **``SegurancaOnline``**: Se possui serviço de proteção online como Antivírus e Antimalware;
* **``BackupOnline``**: Se possui sistema de backup de arquivos em nuvem;
* **``ProtecaodeDispositivo``**: Seguro para proteção contra danos aos dispositivos;
* **``SuporteTecnico``**: Suporte para utilização de dispositivos e serviços;
* **``StreamingTV``**: Serviços de TV sob demanda;
* **``StreamingFilmes``**: Serviços de filmes sob demanada;
* **``Contrato``**: Tipo de plano contrato adotado, podendo ser Anual, Dois Anos ou mensal;
* **``CobrancasemPapel``**: Utilização de boleto por e-mail;
* **``MetododePagamento``**: Forma de pagamento escolhida pelo cliente, podendo ser por correspondência, cartão de crédito ou débito automático;
* **``CobrancasMensais``**: Valores cobrados mensalmente;
* **``CobrancasTotais``**: Somatória dos valores cobrados.

Após esses processos vamos dar início à limpeza dos dados do nosso Data Frame.

In [23]:
# Compreendendo a composição do Data Frame
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ID                     7267 non-null   object 
 1   Evasao                 7267 non-null   object 
 2   Genero                 7267 non-null   object 
 3   Senior                 7267 non-null   int64  
 4   Casado(a)              7267 non-null   object 
 5   Dependentes            7267 non-null   object 
 6   Posse                  7267 non-null   int64  
 7   ServicoTelefonico      7267 non-null   object 
 8   MultiplasLinhas        7267 non-null   object 
 9   ServicodeInternet      7267 non-null   object 
 10  SegurancaOnline        7267 non-null   object 
 11  BackupOnline           7267 non-null   object 
 12  ProtecaodeDispositivo  7267 non-null   object 
 13  SuporteTecnico         7267 non-null   object 
 14  StreamingTV            7267 non-null   object 
 15  Stre

In [24]:
# Verficando os valores inseridos em cada coluna
for coluna in dados.columns:
  print(dados[coluna].value_counts())

ID
0002-ORFBO    1
6614-VBEGU    1
6637-KYRCV    1
6635-MYYYZ    1
6635-CPNUN    1
             ..
3374-TTZTK    1
3374-PZLXD    1
3374-LXDEV    1
3373-YZZYM    1
9995-HOTOH    1
Name: count, Length: 7267, dtype: int64
Evasao
Nao    5174
Sim    1869
        224
Name: count, dtype: int64
Genero
Homem     3675
Mulher    3592
Name: count, dtype: int64
Senior
0    6085
1    1182
Name: count, dtype: int64
Casado(a)
Nao    3749
Sim    3518
Name: count, dtype: int64
Dependentes
Nao    5086
Sim    2181
Name: count, dtype: int64
Posse
1     634
72    369
2     246
3     207
4     185
     ... 
38     60
39     59
44     54
36     50
0      11
Name: count, Length: 73, dtype: int64
ServicoTelefonico
Sim    6560
Nao     707
Name: count, dtype: int64
MultiplasLinhas
Nao                       3495
Sim                       3065
Sem servico telefonico     707
Name: count, dtype: int64
ServicodeInternet
Fibra optica    3198
DSL             2488
Nao             1581
Name: count, dtype: int64
SegurancaO

Verificamos que a coluna Evasão contém 224 valores em branco. Como este é o principal fator que consideraremos para a nossa análise, não há sentido em manter estes 224 valores em nossa análise.

In [25]:
dados['Evasao'].value_counts()

Evasao
Nao    5174
Sim    1869
        224
Name: count, dtype: int64

In [26]:
# Selecionando os dados de interesse para um novo Data Frame
dados_clean = dados[dados['Evasao'] != '']
dados_clean

ID Evasao  Genero  Senior Casado(a) Dependentes  Posse  \
0     0002-ORFBO    Nao  Mulher       0       Sim         Sim      9   
1     0003-MKNFE    Nao   Homem       0       Nao         Nao      9   
2     0004-TLHLJ    Sim   Homem       0       Nao         Nao      4   
3     0011-IGKFF    Sim   Homem       1       Sim         Nao     13   
4     0013-EXCHZ    Sim  Mulher       1       Sim         Nao      3   
...          ...    ...     ...     ...       ...         ...    ...   
7262  9987-LUTYD    Nao  Mulher       0       Nao         Nao     13   
7263  9992-RRAMN    Sim   Homem       0       Sim         Nao     22   
7264  9992-UJOEL    Nao   Homem       0       Nao         Nao      2   
7265  9993-LHIEB    Nao   Homem       0       Sim         Sim     67   
7266  9995-HOTOH    Nao   Homem       0       Sim         Sim     63   

     ServicoTelefonico         MultiplasLinhas ServicodeInternet  ...  \
0                  Sim                     Nao               DSL  ...   
1                  Sim                     Sim               DSL  ...   
2                  Sim                     Nao      Fibra optica  ...   
3                  Sim                     Nao      Fibra optica  ...   
4                  Sim                     Nao      Fibra optica  ...   
...                ...                     ...               ...  ...   
7262               Sim                     Nao               DSL  ...   
7263               Sim                     Sim      Fibra optica  ...   
7264               Sim                     Nao               DSL  ...   
7265               Sim                     Nao               DSL  ...   
7266               Nao  Sem servico telefonico               DSL  ...   

     BackupOnline ProtecaodeDispositivo SuporteTecnico StreamingTV  \
0             Sim                   Nao            Sim         Sim   
1             Nao                   Nao            Nao         Nao   
2             Nao                   Sim            Nao         Nao   
3             Sim                   Sim            Nao         Sim   
4             Nao                   Nao            Sim         Sim   
...           ...                   ...            ...         ...   
7262          Nao                   Nao            Sim         Nao   
7263          Nao                   Nao            Nao         Nao   
7264          Sim                   Nao            Nao         Nao   
7265          Nao                   Sim            Sim         Nao   
7266          Sim                   Sim            Nao         Sim   

     StreamingFilmes   Contrato CobrancasemPapel MetododePagamento  \
0                Nao      Anual              Sim   Correspondencia   
1                Sim     Mensal              Nao   Correspondencia   
2                Nao     Mensal              Sim        Eletronico   
3                Sim     Mensal              Sim        Eletronico   
4                Nao     Mensal              Sim   Correspondencia   
...              ...        ...              ...               ...   
7262             Nao      Anual              Nao   Correspondencia   
7263             Sim     Mensal              Sim        Eletronico   
7264             Nao     Mensal              Sim   Correspondencia   
7265             Sim  Dois anos              Nao   Correspondencia   
7266             Sim  Dois anos              Nao        Eletronico   

     CobrancasMensais  CobrancasTotais  
0               65.60            593.3  
1               59.90            542.4  
2               73.90           280.85  
3               98.00          1237.85  
4               83.90            267.4  
...               ...              ...  
7262            55.15            742.9  
7263            85.10           1873.7  
7264            50.30            92.75  
7265            67.85          4627.65  
7266            59.00           3707.6  

[7043 rows x 21 columns]

In [27]:
# Resetando o Index
dados_clean.reset_index(drop = True, inplace = True)
dados_clean

ID Evasao  Genero  Senior Casado(a) Dependentes  Posse  \
0     0002-ORFBO    Nao  Mulher       0       Sim         Sim      9   
1     0003-MKNFE    Nao   Homem       0       Nao         Nao      9   
2     0004-TLHLJ    Sim   Homem       0       Nao         Nao      4   
3     0011-IGKFF    Sim   Homem       1       Sim         Nao     13   
4     0013-EXCHZ    Sim  Mulher       1       Sim         Nao      3   
...          ...    ...     ...     ...       ...         ...    ...   
7038  9987-LUTYD    Nao  Mulher       0       Nao         Nao     13   
7039  9992-RRAMN    Sim   Homem       0       Sim         Nao     22   
7040  9992-UJOEL    Nao   Homem       0       Nao         Nao      2   
7041  9993-LHIEB    Nao   Homem       0       Sim         Sim     67   
7042  9995-HOTOH    Nao   Homem       0       Sim         Sim     63   

     ServicoTelefonico         MultiplasLinhas ServicodeInternet  ...  \
0                  Sim                     Nao               DSL  ...   
1                  Sim                     Sim               DSL  ...   
2                  Sim                     Nao      Fibra optica  ...   
3                  Sim                     Nao      Fibra optica  ...   
4                  Sim                     Nao      Fibra optica  ...   
...                ...                     ...               ...  ...   
7038               Sim                     Nao               DSL  ...   
7039               Sim                     Sim      Fibra optica  ...   
7040               Sim                     Nao               DSL  ...   
7041               Sim                     Nao               DSL  ...   
7042               Nao  Sem servico telefonico               DSL  ...   

     BackupOnline ProtecaodeDispositivo SuporteTecnico StreamingTV  \
0             Sim                   Nao            Sim         Sim   
1             Nao                   Nao            Nao         Nao   
2             Nao                   Sim            Nao         Nao   
3             Sim                   Sim            Nao         Sim   
4             Nao                   Nao            Sim         Sim   
...           ...                   ...            ...         ...   
7038          Nao                   Nao            Sim         Nao   
7039          Nao                   Nao            Nao         Nao   
7040          Sim                   Nao            Nao         Nao   
7041          Nao                   Sim            Sim         Nao   
7042          Sim                   Sim            Nao         Sim   

     StreamingFilmes   Contrato CobrancasemPapel MetododePagamento  \
0                Nao      Anual              Sim   Correspondencia   
1                Sim     Mensal              Nao   Correspondencia   
2                Nao     Mensal              Sim        Eletronico   
3                Sim     Mensal              Sim        Eletronico   
4                Nao     Mensal              Sim   Correspondencia   
...              ...        ...              ...               ...   
7038             Nao      Anual              Nao   Correspondencia   
7039             Sim     Mensal              Sim        Eletronico   
7040             Nao     Mensal              Sim   Correspondencia   
7041             Sim  Dois anos              Nao   Correspondencia   
7042             Sim  Dois anos              Nao        Eletronico   

     CobrancasMensais  CobrancasTotais  
0               65.60            593.3  
1               59.90            542.4  
2               73.90           280.85  
3               98.00          1237.85  
4               83.90            267.4  
...               ...              ...  
7038            55.15            742.9  
7039            85.10           1873.7  
7040            50.30            92.75  
7041            67.85          4627.65  
7042            59.00           3707.6  

[7043 rows x 21 columns]

In [28]:
# Verificando se a seleção foi realizada com êxito
dados_clean['Evasao'].value_counts()

Evasao
Nao    5174
Sim    1869
Name: count, dtype: int64

Verificamos que também há 11 valores em branco na coluna CobrancasTotais, contudo esses clientes ainda podem servir de algum valor para nossa análise. Vamos verificar o que ocorre com esta seleção específica.

In [29]:
# Seleciecionando os 11 casos especificados
dados_clean[dados_clean['CobrancasTotais'] == ' '][['Posse', 'CobrancasMensais', 'CobrancasTotais']]

Posse  CobrancasMensais CobrancasTotais
945       0             56.05                
1731      0             20.00                
1906      0             61.90                
2025      0             19.70                
2176      0             20.25                
2250      0             25.35                
2855      0             73.35                
3052      0             25.75                
3118      0             52.55                
4054      0             80.85                
5433      0             19.85

Como observado, não são casos em que houve evasão, mas novos clientes os quais tiveram apenas uma cobrança. Isso é verificado pelo tempo de posse apresentado pelo cliente que consta como zero. Não devemos apagar estes dados, sendo assim devemos duplicar o valor do único mês cobrado para cobranças totais.

In [30]:
# Alterando os valores
sem_total = dados_clean[dados_clean['CobrancasTotais'] == ' '].index
dados_clean.loc[sem_total, 'CobrancasTotais'] = dados_clean.loc[sem_total, 'CobrancasMensais']

In [31]:
# Verificando se ainda temos valores em branco em CobrancasTotais
dados_clean[dados_clean['CobrancasTotais'] == ' '][['Posse', 'CobrancasMensais', 'CobrancasTotais']]

Empty DataFrame
Columns: [Posse, CobrancasMensais, CobrancasTotais]
Index: []

In [32]:
# Verificando os clientes com tempo de posse igual a zero
dados_clean[dados_clean['Posse'] == 0][['Posse', 'CobrancasMensais', 'CobrancasTotais']]

Posse  CobrancasMensais CobrancasTotais
945       0             56.05           56.05
1731      0             20.00            20.0
1906      0             61.90            61.9
2025      0             19.70            19.7
2176      0             20.25           20.25
2250      0             25.35           25.35
2855      0             73.35           73.35
3052      0             25.75           25.75
3118      0             52.55           52.55
4054      0             80.85           80.85
5433      0             19.85           19.85

# Questões de negócio

Após realizarmos EDA, pudemos compreender do que se tratam as Features e tratar os dados ausentes do nosso dataset, sendo assim, podemos estabelecer questões de negócio que norteiem o Projeto do Modelo Preditivo de Churn. As questões que melhor auxiliam os objetivos estabelecido são:

1.	Quais são os principais fatores que influenciam a evasão de clientes?
2.	Qual é o perfil demográfico dos clientes que mais evadem?
3.	Como a duração do contrato (mensal, anual, dois anos) afeta a taxa de churn?
4.	A adesão a serviços adicionais tem impacto na retenção de clientes?
5.	Clientes com múltiplas linhas telefônicas são menos propensos a evadir?
6.	O método de pagamento influencia a taxa de churn?
7.	A existência de cobranças em papel está associada a uma maior taxa de churn?
8.	Qual é o impacto das cobranças mensais e totais no churn?
9.	A duração do serviço (tempo de posse) afeta a probabilidade de churn?
10.	Como a qualidade do serviço de internet afeta a evasão de clientes?

Essas questões de negócio nos fornecem uma base sólida para a análise e modelagem preditiva, o que irá auxiliar a identificar padrões e construir um modelo robusto para a Previsão de Churn.

In [33]:
dados_clean.to_csv('aluravoz_clean.csv', encoding = 'utf-8', index = False)